### DISCUSSION TITLE
Decrypting the hash of NAT_DETECTION_SOURCE_IP and NAT_DETECTION_DESTINATION_IP?

### DISCUSSION THREAD

**Discussion by burstyyy** [strongswan/strongswan#2347] [Q&A] [answered]
Hi there,

I was wondering if there's any particular way to decrypt the hash of NAT_DETECTION_SOURCE_IP and NAT_DETECTION_DESTINATION_IP sent in an IKE_SA_INIT packet in Strongswan logging (or after the fact if you have the decryption keys?)

We often run into situations where it'd be helpful to have the decrypted and encrypted output of this, mostly to ensure that the remote peer is doing the right thing.

If there's any debug that can be added, that'd be incredibly helpful!

Thanks in advance!

**Comment by tobiasbrunner** [comment]
You already said it in your question, these are hashes. So you can't directly reverse them. They are the SHA-1 hash of the two SPIs, an IP address and a UDP port. You could possibly brute force this to determine the IP address/port if you know the SPIs and hash value (at least for IPv4).

> We often run into situations where it'd be helpful to have the decrypted and encrypted output of this, mostly to ensure that the remote peer is doing the right thing.

How so? What exactly is the issue?

> If there's any debug that can be added, that'd be incredibly helpful!

strongSwan does log some details on its own calculation of the hashes on level 3 in _ike_. But that won't really help you if the peer does something wrong, there simply will be a mismatch of the calculated and received hashes.

**Reply by burstyyy** [reply]
Thanks for the quick reply Tobias!

> So you can't directly reverse them. They are the SHA-1 hash of the two SPIs, an IP address and a UDP port.

The inspiration for this post was from Junipers SRX firewalls - Junipers implementation has a way to see the decrypted values before and after, and this was kind of the inspiration for this post - https://www.blackhole-networks.com/IKE_Modes/ikev2-initial.html (search for NAT_DETECTION and look at the bottom example). 

Since the values are also salted, I assumed that the Strongswan (and any implementation) needs to have a way to unsalt and unhash it to determine if there is a NAT situation (otherwise, how could Strongswan tell what the remote peer was sending in its NAT_DETECTION_SOURCE_IP value?), but I'd love to understand more if that's not the case!

> How so? What exactly is the issue?

I've had issues with other vendors where they would either not identify a NAT situation when there is one, or identify a NAT situation when there isn't one, and whilst we can assume it is another vendor's problem, it's difficult to say without being able to see what the decrypted values of the IP/port/SPIs are from both peers. 

I don't have a specific example to share right now, however - it's more of a curiosity I've had over time, and having an objective response to some of those concerns when those situations arise would be great!

**Reply by tobiasbrunner** [reply]
> Junipers implementation has a way to see the decrypted values before and after, and this was kind of the inspiration for this post

As I said, strongSwan can print the same thing for the NAT-D hashes it calculates. It won't for the hashes it receives (neither will Juniper).

> Since the values are also salted, I assumed that the Strongswan (and any implementation) needs to have a way to unsalt and unhash it to determine if there is a NAT situation (otherwise, how could Strongswan tell what the remote peer was sending in its NAT_DETECTION_SOURCE_IP value?), but I'd love to understand more if that's not the case!

No, it does not "unhash" the value (there is no such thing ;-). It simply calculates the expected value (using the SPIs, IP and port from the received packet) and compares it to the value in the received NAT-D payload the peer calculated. If they differ, there is a NAT in between because the peer saw a different IP/port when it sent the packet (or it is faking a NAT situation by sending a random value).

> I've had issues with other vendors where they would either not identify a NAT situation when there is one, or identify a NAT situation when there isn't one

As mentioned above, the latter can happen if the peer fakes a NAT situation, e.g. to force UDP encapsulation (strongSwan can do so as well). The former really can only happen if no NAT-D payloads are sent at all (i.e. NAT-T it not supported or disabled by the peer), or maybe if the peer had a setting to configure a public IP to use when generating the NAT-D hash to avoid that a NAT is detected in static NAT situations (e.g. port forwarding for a VPN gateway).

**Reply by burstyyy** [reply]
Hey Tobias, thanks for the reply!

> As I said, strongSwan can print the same thing for the NAT-D hashes it calculates. It won't for the hashes it receives (neither will Juniper).

Thanks for clarifying - so the output from the SRX in that article is actually calculating what the hash should be for themselves and for the remote peer, and then matching that with what it receives from the remote peer. Got it! I had a look at the IKE=3 value and can see that you mean - I see the precalculated src/dst_hashs and the src/dst hashes received, I must have skimmed it because I was looking out wording similar to the SRX.

> It simply calculates the expected value (using the SPIs, IP and port from the received packet) and compares it to the value in the received NAT-D payload the peer calculated.

It was my original understanding that these hashes are also salted, in which case just re-hashing the expected value on our end still wouldn't match (and you can see this in packet captures, where the values don't match, even in a no NAT scenario) - am I wrong in my understanding there, and they're actually not salted, and just hashed?

> The former really can only happen if no NAT-D payloads are sent at all (i.e. NAT-T it not supported or disabled by the peer), or maybe if the peer had a setting to configure a public IP to use when generating the NAT-D hash to avoid that a NAT is detected in static NAT situations (e.g. port forwarding for a VPN gateway).

To give an example I remember recently that brought up the idea of "checking the hashes" - Fortinet has odd behaviour where when they have NAT-Traversal enabled, the tunnel will form with UDP encapsulated ESP, but after IKE_SA reauthentication, the tunnel then forms without UDP encapsulation (just ESP), however, Fortinet still sends UDP encapsulated ESP (which as we know is allowed by the standard but Strongswan doesn't support) which we simply drop.

I was just covering all of our bases on our side to try and understand why they're switching encapsulation method after reauthenticating, but it's likely just something to do with their implementation (to try understand how they send mismatching hashes originally, but not during reauth).

I appreciate all of your time with this, thank you!

**Reply by tobiasbrunner** [reply, marked as answer]
> It was my original understanding that these hashes are also salted, in which case just re-hashing the expected value on our end still wouldn't match (and you can see this in packet captures, where the values don't match, even in a no NAT scenario) - am I wrong in my understanding there, and they're actually not salted, and just hashed?

There is no specific salt, but I guess you could consider the SPIs salts (these are 64-bit values, often random, however, in the IKE_SA_INIT request only the initiator SPI is set), which make the hashes of the same IP address and port different for different IKE_SAs. But as I mentioned, an implementation might just send a random value to fake a NAT situation to force UDP-encapsulation.

> To give an example I remember recently that brought up the idea of "checking the hashes" - Fortinet has odd behaviour where when they have NAT-Traversal enabled, the tunnel will form with UDP encapsulated ESP, but after IKE_SA reauthentication, the tunnel then forms without UDP encapsulation (just ESP), however, Fortinet still sends UDP encapsulated ESP (which as we know is allowed by the standard but Strongswan doesn't support) which we simply drop.

Yeah, the Linux kernel currently doesn't support that. But it also doesn't really make sense, if they want to force UDP-encapsulation, they could just fake a NAT (you can do so from your end with such a peer if necessary).

> I was just covering all of our bases on our side to try and understand why they're switching encapsulation method after reauthenticating, but it's likely just something to do with their implementation (to try understand how they send mismatching hashes originally, but not during reauth).

Sounds like a bug.

**Reply by burstyyy** [reply]
I appreciate all of your replies, Tobias!
